cpp/src/arrow/dataset/dataset_internal.h

```diff
@@ -204,5 +204,35 @@ arrow::Result<std::shared_ptr<T>> GetFragmentScanOptions(
   return internal::checked_pointer_cast<T>(source);
 }
 
+class FragmentDataset : public Dataset {
```

Comment:
Should this be a base type of `InMemoryDataset`?

Reply:
Whether or not the component fragments are in memory is up to the fragments, so I don't think it's appropriate to inherit that here

Reply:
Right, sorry, I meant the other way around.  `InMemoryDataset : public FragmentDataset`.

Reply:
Added https://issues.apache.org/jira/browse/ARROW-12945 to track this improvement.